config.py: Support per-user configuration file (draft)

fgyanz opened the pull request:
Add support for a per-user configuration file as the main method of configuring klp-build instead of relying on user-set environment variables.

The new implementation what basically does is:
Upon startup, check if `$HOME/.config/klp-build/config` already exists. If it does, load the configuration. Otherwise, create a default configuration file and same with the files and directories specified in the conf.

The default paths for now are:
work directory -> `$HOME/klp-build/patches`
data directory -> `$HOME/klp-build/data`

fgyanz committed:
config: Support per-user configuration file

Add support for a per-user configuration file as the main
method of configuring klp-build instead of relying on
user-defined environment variables.

The conf file is subdivided into sections, and each
one of them supports a set of variables that can be defined
by the user, following the common "<key>=<value>" format.

Right now the only supported section is "Paths", which
encapsulates the directories and files paths needed
by klp-build. E.g:
```
[Paths]
work_dir = myworkdir/
data_dir = mydatadir/
```

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

marcosps commented:
Overall I liked the approach. But I think that we could have example of the config file in the commit message, so we could easily understand the end goal of it. Also, I'm adding a comments in the commit directly.

marcosps commented on klpbuild/config.py:
```diff
@@ -29,13 +30,19 @@
-        work = Path(work_dir)
-        if not work.is_dir():
-            raise ValueError("Work dir should be a directory")
+        self.user_conf_file = Path(home, ".config/klp-build/config")
```
I believe that we can expect users to have HOME defined? I mean, the majority of people use the most common distros, and I have never saw anyone without HOME being defined :)

marcosps replied:
It's still the same as the older version. You can use pathlib's Path().home() to get the home, and I think that we don't need to check if it is defined or not...

marcosps commented on klpbuild/config.py:
```diff
@@ -61,20 +68,54 @@ def __init__(self, lp_name, lp_filter, kdir=False, data_dir=None, skips="", work
         logging.basicConfig(level=logging.INFO, format="%(message)s")
 
+    def setup_user_env(self, basedir):
+        basedir = Path(basedir, "klp-build")
```
At this point we are overriding the argument value? For me it's confusing =/

fgyanz replied:
Actually, it's more like appending. So, initially `basedir` is `$HOME` , and after this line `basedir` becomes `$HOME/klp-build`. I'll try to rewrite this part to avoid misleading people.

marcosps commented on klpbuild/config.py:
```diff
@@ -61,20 +68,54 @@ def __init__(self, lp_name, lp_filter, kdir=False, data_dir=None, skips="", work
 
+    def setup_user_env(self, basedir):
+        basedir = Path(basedir, "klp-build")
+        workdir = Path(basedir, "patches")
```
I would prefer having the directory "livepatches" over patches, since it's common to have a "patches" directory. Stating as "livepatches" directory seems more objective.

marcosps commented on klpbuild/config.py:
```diff
@@ -61,20 +68,54 @@ def __init__(self, lp_name, lp_filter, kdir=False, data_dir=None, skips="", work
+        p = Path(self.user_conf['Paths'][entry])
+        if not p.exists():
+            raise ValueError(f"'{p}' file or directory not found")
+        if isdir is True and not p.is_dir():
```
You can check isdir directly, like

```
if isdir:
```

marcosps commented:
Btw, we should show messages on the screen when we create a new config or load the existing one. I believe that this would inform the user of the tool where they can change these values to point the LP creating or data placement into a different directory.

(Also I'm planning to support creating a LP using /lib/modules and /boot of the host machine, so I would need special values on the data-dir config for it to work properly, so showing the user where to look at the config is useful)

fgyanz committed:
config: Support 'kernel-source' in config file

Allow setting the 'kernel-source' path in the per-user
configuration file as an alternative to "KLP_KERNEL_SOURCE"
environment variable.
```
[Paths]
kernel_src_dir = mykernelsrc/
```

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
config: Support 'ccp-pol-scripts' in config file

Allow setting the 'ccp-pol-scripts' path in the per-user
configuration file as an alternative to "KLP_CCP_POL_PATH"
environment variable.
```
[Paths]
ccp_pol_dir = myccppol/
```

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

marcosps commented on klpbuild/config.py:
```diff
@@ -80,8 +80,10 @@ def setup_user_env(self, basedir):
         config['Paths'] = {'work_dir': workdir,
-                           'data_dir': datadir}
+                           'data_dir': datadir,
+                           '#kernel_src_dir': 'kernel-src/ (e.g. https://github.com/SUSE/kernel-source)'}
```
I think that kernel-source doesn't contain the CVE branches that we have on kerncvs, so we may remove it. Also, IIRC, current we don't require this value to be defined, so I think that we can relax the check for this setting.

Per my tests it fails if kernel_src_dir is not set, or if it's created by klpbuild but it's commented.

marcosps commented on klpbuild/config.py:
```diff
@@ -29,13 +30,19 @@
+            logging.warning(f"Warning: user configuration file not found")
+            # If there's no configuration file assume fresh install.
+            # Prepare the system with a default environment and conf.
+            self.setup_user_env(Path(home, "klp-build"))
```
Here I think that we can rename the directory to ``klp``, since klp-build can be confused with the project name. We were using ``kgr`` for historic reasons, but currently we use ``klp`` everywhere.

fgyanz committed:
klp-ccp: Remove KLP_CCP_PATH env variable

From now on KLP_CCP_PATH won't be allowed anymore as a way
to find klp-ccp's binary.

The binary will be assumed to be in one of the directories
in PATH env variable (E.g. /bin/, /usr/bin...)

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
config: Support 'patches-tests' in config file

Allow setting the 'patches-tests' path in the per-user
configuration file instead of having hardcoded paths.
```
[Paths]
patches_tests_dir = mytestsdir/
```

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
config: Support 'kgraft-patches' in config file

Allow setting the 'kgraft-patches' path in the per-user
configuration file instead of using hardcoded paths.
```
[Paths]
patches_dir = mypatchesdir/
```

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>
